moonraker/components/klipper_estimator.py

```diff
@@ -0,0 +1,101 @@
+        self.gc_path = self.file_manager.get_directory()
+        hostname = self.server.get_host_info()["hostname"]
+        port = self.server.get_host_info()["port"]
+        self.url = f"http://{hostname}:{port}/"
+        self.server.register_event_handler(
+            "file_manager:filelist_changed",
+            self._handle_filelist_changed
+        )
+        if self.download:
+            asyncio.create_task(self._download())
+
+    async def _download(self) -> None:
+        client: HttpClient = self.server.lookup_component("http_client")
+        response = await client.github_api_request(
+            f"repos/Annex-Engineering/klipper_estimator/releases/{self.tag}")
+        response.raise_for_status(
+            "Failed to get latest Klipper Estimator info.")
+        data = cast(dict, response.json())
+        tag_name = data["tag_name"]
+        try:
+            current_version = open(f"{self.ke_exec}.version", "r").read()
+        except Exception as e:
+            logging.info("Cant determine current version")
+            logging.info(e)
+            current_version = "9999"
+        if current_version != tag_name or not os.path.isfile(self.ke_exec):
+            for asset in data["assets"]:
+                if asset["name"] == self.asset_name:
+                    down_url = asset["browser_download_url"]
+                    content_type = asset["content_type"]
+                    logging.info(
+                        f"Downloading {self.asset_name} from {down_url}")
+                    try:
+                        b = await client.get_file(down_url, content_type)
+                        f = open(self.ke_exec, "wb")
+                        f.write(b)
+                        f.close()
+                        os.chmod(self.ke_exec,
+                                 os.stat(self.ke_exec).st_mode | stat.S_IEXEC)
+                    except Exception as e:
+                        logging.error(e)
+                    logging.info("Done")
+                    file = open(f"{self.ke_exec}.version", "w")
+                    file.write(tag_name)
+                    file.close()
+        else:
+            logging.info("klipper-estimator already latest")
+
+    async def _handle_filelist_changed(self, event) -> None:
+        action = event["action"]
+        path = event["item"]["path"]
+        if action in ["create_file", "modify_file"]:
+            if path.lower().endswith(".gcode"):
+                full_path = f"{self.gc_path}/{path}"
+                logging.info(f"Running klipper-estimator in {path}")
+                await self._run_estimator(full_path)
+
+    async def _run_estimator(self, path) -> None:
+        path = path.replace("\"", "\\\"")
+        cmd = " ".join(
+            [self.ke_exec, "--config_moonraker_url", self.url, "post-process",
+             f"\"{path}\""])
+        logging.info(f"Running {cmd}")
+        timeout = 10.
```

Comment:
During my experimentations, I've noticed that the rpi4 is often not able to process the gcode files in 10 seconds. I've now increased it for me to 120s inside the code. Not sure if it would be better to make it configurable from the outside.

Reply:
I think it's worth adding as a config, I'll add it.